### PR TITLE
Feature: Support Event Filtering for Device Mode Destinations

### PR DESCRIPTION
**Pull request by saikumarrs**
## Description of the change

Added a feature to support event filtering for device mode destinations.
- Event Filtering feature allows customers to filter events at the SDK level. This feature can be enabled from the control plane. There are three options, only one can be selected :
    - Disable: Disable the feature.
    - Blacklist: Black-listed events will be filtered out.
    - Whitelist: Only white-listed events will flow through.

- Selected option will be supplied in the parameter `eventFilteringOption`. Black/White listed Events to be supplied as an array parameter in these`blacklistedEvents`, `whitelistedEvents` in the destination config object.
- Also, `indes.js` file ganarated for NPM package added in the `.gitignore`.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

N/A

### Development

- [x] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/358)

### CODE REVIEW

**Review by MoumitaM — DISMISSED** [review]

**Review by MoumitaM — APPROVED** [review]

File: dist/rudder-sdk-js/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "rudder-sdk-js",
-  "version": "1.2.15",
+  "version": "1.2.18",
```
Comment (pallabmaiti) [inline dist/rudder-sdk-js/package.json:3]: Should not we change to 1.3.0 as event filtering is a new feature?

Comment (saikumarrs) [reply dist/rudder-sdk-js/package.json:3]: Yeah, we should. We usually bump the version once the changes are merged to production.

**Comment by saikumarrs** [comment]
[Non-blocking]
PR description needs to be updated.

File: .gitignore [hunk 2]
```diff
@@ -17,5 +17,6 @@ build.sh
 publish.sh
 dist/*
 !dist/rudder-sdk-js/*
+dist/rudder-sdk-js/index.js
```
Comment (saikumarrs) [inline .gitignore:20]: [Non-blocking]
PR description needs to be updated.

Comment (MoumitaM) [reply .gitignore:20]: done

File: analytics.js [hunk 3]
```diff
@@ -669,6 +678,66 @@ class Analytics {
...
+            }
+          default:
+            return false;
+        }
+      } else {
+        return false;
+      }
+    }
+    return false;
+  }
```
Comment (saikumarrs) [inline analytics.js:739]: Perform validation in the beginning and return from there itself. Makes this logic less nested and readable.

Comment (MoumitaM) [reply analytics.js]: done

File: analytics.js [hunk 4]
```diff
@@ -679,7 +679,9 @@ class Analytics {
   }
 
   IsEventBlackListed(eventName, intgName) {
-    if (typeof eventName === "string" && eventName) {
+    if (!eventName || !(typeof eventName === "string")) {
+      return false;
+    } else {
```
Comment (saikumarrs) [inline analytics.js:684]: There is no need for `else` blocks.

```
if (<validation>) {
  // some more logic
  return <boolean>;
}

// continue with more logic
```

**Review by pallabmaiti — APPROVED** [review]

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**